pkg/retry/bigquery.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package retry
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"cloud.google.com/go/bigquery"
+	"github.com/abcxyz/pkg/logging"
+	"go.uber.org/zap"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+// BigQuery provides a client and dataset identifiers.
+type BigQuery struct {
+	projectID string
+	datasetID string
+	client    *bigquery.Client
+	logger    *zap.SugaredLogger
+}
+
+// CheckpointEntry is the shape of an entry to the checkpoint table.
+type CheckpointEntry struct {
+	deliveryID string
+	createdAt  string
+}
+
+// NewBigQuery creates a new instance of a BigQuery client.
+func NewBigQuery(ctx context.Context, projectID, datasetID string, opts ...option.ClientOption) (*BigQuery, error) {
+	client, err := bigquery.NewClient(ctx, projectID, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new bigquery client: %w", err)
+	}
+
+	return &BigQuery{
+		projectID: projectID,
+		datasetID: datasetID,
+		client:    client,
+		logger:    logging.FromContext(ctx).Named("bigquery"),
+	}, nil
+}
+
+// Shutdown releases any resources held by the BigQuery client.
+func (bq *BigQuery) Shutdown() error {
```

Comment:
Nit: this should probably be named "Close"

Reply:
ack, pubsub also has a Shutdown, should we be consistent and switch to Close there?

Reply:
The value of `Close() error` is that it satisfies the `io.Closer` interface.

Reply:
ah understood